src/core/pointcloud/expression/qgspointcloudexpressionnodeimpl.cpp

```diff
@@ -0,0 +1,801 @@
+  return copy;
+}
+
+bool QgsPointCloudExpressionNodeUnaryOperator::isStatic( QgsPointCloudExpression *parent, const QgsPointCloudBlock *block ) const
+{
+  return mOperand->isStatic( parent, block );
+}
+
+QString QgsPointCloudExpressionNodeUnaryOperator::text() const
+{
+  return UNARY_OPERATOR_TEXT[mOp];
+}
+
+//
+
+double QgsPointCloudExpressionNodeBinaryOperator::evalNode( QgsPointCloudExpression *parent, int p )
+{
+  double vL = mOpLeft->eval( parent, p );
+  ENSURE_NO_EVAL_ERROR
+
+  if ( mOp == boAnd || mOp == boOr )
+  {
+    if ( mOp == boAnd && vL == 0. )
+      return 0.;  // shortcut -- no need to evaluate right-hand side
+    if ( mOp == boOr && vL != 0. )
+      return 1.;  // shortcut -- no need to evaluate right-hand side
+  }
+
+  double vR = mOpRight->eval( parent, p );
+  ENSURE_NO_EVAL_ERROR
+
+  switch ( mOp )
+  {
+    case boPlus:
+      return vL + vR;
+    case boMinus:
+      return vL - vR;
+    case boMul:
+      return vL * vR;
+    case boDiv:
+      if ( vR == 0. )
+        return std::numeric_limits<double>::quiet_NaN();
+      return vL / vR;
+    case boMod:
+      return std::fmod( vL, vR );
+    case boIntDiv:
+      return std::floor( vL / vR );
+    case boPow:
+      return std::pow( vL, vR );
+    case boAnd:
+      // vL is already checked and is true (not 0.)
+      return vR == 0. ? 0. : 1.;
+    case boOr:
+      // vL is already checked and is false (0.)
+      return vR == 0. ? 0. : 1.;
+
+    case boEQ:
+    case boNE:
+    case boLT:
+    case boGT:
+    case boLE:
+    case boGE:
+      return compare( vL - vR ) ? 1. : 0.;
+
+    case boNotImplemented:
+      return std::numeric_limits<double>::quiet_NaN();
+  }
+  Q_ASSERT( false );
+  return std::numeric_limits<double>::quiet_NaN();
+}
+
+bool QgsPointCloudExpressionNodeBinaryOperator::compare( double diff )
+{
+  switch ( mOp )
+  {
+    case boEQ:
+      return diff == 0.0;
+    case boNE:
+      return diff != 0.0;
+    case boLT:
+      return diff < 0;
+    case boGT:
+      return diff > 0;
+    case boLE:
+      return diff <= 0;
+    case boGE:
+      return diff >= 0;
+    default:
+      Q_ASSERT( false );
+      return std::numeric_limits<double>::quiet_NaN();
+  }
+}
+
+qlonglong QgsPointCloudExpressionNodeBinaryOperator::computeInt( qlonglong x, qlonglong y )
+{
+  switch ( mOp )
+  {
+    case boPlus:
+      return x + y;
+    case boMinus:
+      return x - y;
+    case boMul:
+      return x * y;
+    case boDiv:
+      return x / y;
+    case boMod:
+      return x % y;
+    default:
+      Q_ASSERT( false );
+      return 0;
+  }
+}
+
+double QgsPointCloudExpressionNodeBinaryOperator::computeDouble( double x, double y )
+{
+  switch ( mOp )
+  {
+    case boPlus:
+      return x + y;
+    case boMinus:
+      return x - y;
+    case boMul:
+      return x * y;
+    case boDiv:
+      return x / y;
+    case boMod:
+      return std::fmod( x, y );
+    default:
+      Q_ASSERT( false );
+      return 0;
+  }
+}
```

Comment:
it seems to me that computeInt() and computeDouble() are not used anywhere, all evaluation is done directly in evalNode()

Reply:
indeed!